### ISSUE TITLE
Starting off with different models across ranks and FSDP doesn't synchronise

### ISSUE THREAD

**Issue by BadrYoubiIdrissi** [pytorch/torchtitan#166]
https://github.com/pytorch/torchtrain/blob/8dd5798241490c5f532e822e9f9c1d30e0fba0df/train.py#L155-L159

Hey there ! With @mathuvu we ran into a sneaky bug in our codebase that broke loss evolution parity with xlformers. 

Basically we didn't seed torch before initialising the model so each data parallel rank got a different model. We used pytorch's FSDP1 and realised that it never synced the model weights at startup so we ended up optimising N different models but with the same gradient. This gave us much higher loss than xlformer and it also stagnated much earlier. But surprisingly the model was still training and the loss still went down !

I quickly searched in this codebase for torch manual seeding and didn't find anything so i wanted to raise this issue in case i can spare some frustrating debugging!

After fixing this and a few differences in initialisation we now match the loss of xlformer exactly at least in the no_shard setting on small models when starting with the same init.

**Comment by awgu** [comment]
I am going to close this issue for now since this is related to FSDP1 `NO_SHARD`, not FSDP2.
